File: ios/Classes/Data/Repository/TaskRepositoryImpl.swift
import Foundation

/// Repository implementation for working with HTTP request tasks.
///
/// Uses TaskQueueManager to manage the task queue,
/// which helps avoid hangs when registering a large number of tasks.
actor TaskRepositoryImpl: TaskRepository {
    private let fileStorage: FileStorageDataSource
    private let urlSession: URLSessionDataSource
    private let queueManager: TaskQueueManager
    
    // Set to track requests that are currently being created
    private var creatingTasks = Set<String>()
    
    init() {
        self.fileStorage = FileStorageDataSource()
        self.urlSession = URLSessionDataSource()
        self.queueManager = TaskQueueManager.shared
        
        // Set callback for executing tasks from the queue
        Task {
            await queueManager.setExecuteCallback { [weak self] requestId in
                await self?.executeQueuedTask(requestId: requestId)
            }
            
            // Process restored queue
            await queueManager.processQueue()
        }
    }
    
    func createTask(request: HttpRequest) async throws -> TaskInfo {
        let requestId = request.requestId ?? RequestMapper.generateRequestId()
        
        // Race-condition protection: use the actor for synchronization
        // Check whether a request with such requestId is already being created
        if creatingTasks.contains(requestId) {
            // Request is already being created in another context; wait and return the existing one
            try await Task.sleep(nanoseconds: 100_000_000) // 100ms
            if let existingTask = try await getTaskInfo(requestId: requestId) {
                return existingTask
            }
            throw NSError(domain: "TaskRepositoryImpl", code: -2, userInfo: [NSLocalizedDescriptionKey: "Failed to get task info after creation attempt"])
        }
        
        // Check whether a request with such requestId already exists
        if let existingTask = try await getTaskInfo(requestId: requestId) {
            // Request already exists
            // Check whether it is in the queue or active
            if await queueManager.isTaskPendingOrActive(requestId) {
                return existingTask
            }
            
            // Check request status
            let status = existingTask.status
            if status == .inProgress {
                // Request is already executing; enqueue it
                _ = await queueManager.enqueue(requestId)
                return existingTask
            } else if status == .completed || status == .failed {
                // Request is already finished; return existing
                return existingTask
            }
        }
        
        // Mark that the request is being created
        creatingTasks.insert(requestId)
        
        defer {
            // Remove from the set of requests being created
            creatingTasks.remove(requestId)
        }
        
        let registrationDate = Int64(Date().timeIntervalSince1970 * 1000)
        
        // Save request to file
        let taskInfo = try fileStorage.saveRequest(request: request, requestId: requestId, registrationDate: registrationDate)
        
        // Add task to the queue (instead of starting it directly)
        // TaskQueueManager itself will decide when to start the task
        _ = await queueManager.enqueue(requestId)
        
        return taskInfo
    }
    
    /// Executes a task from the queue
    private func executeQueuedTask(requestId: String) async {
        guard let taskInfo = fileStorage.loadTaskInfo(requestId: requestId) else {
            print("[TaskRepositoryImpl] Task info not found for: \(requestId)")
            await queueManager.onTaskCompleted(requestId)
            return
        }
        
        // Load request data from file
        guard let requestData = fileStorage.loadRequestData(requestId: requestId) else {
            print("[TaskRepositoryImpl] Request data not found for: \(requestId)")
            await queueManager.onTaskCompleted(requestId)
            return
        }
        
        guard let url = URL(string: requestData.url) else {
            print("[TaskRepositoryImpl] Invalid URL for: \(requestId)")
            await queueManager.onTaskCompleted(requestId)
            return
        }
        
        // Execute request
        do {
            try await urlSession.executeRequest(
                requestId: requestId,
                url: url,
                method: requestData.method,
                headers: requestData.headers,
                body: requestData.body,
                queryParameters: requestData.queryParameters,
                timeout: requestData.timeout ?? 120,
                multipartFields: requestData.multipartFields,
                multipartFiles: requestData.multipartFiles,
                retries: requestData.retries ?? 0,
                fileStorage: fileStorage,
                onCompleted: { [weak self] in
                    Task {
                        await self?.queueManager.onTaskCompleted(requestId)
                    }
                }
            )
        } catch {
            print("[TaskRepositoryImpl] Error executing request \(requestId): \(error)")
            await queueManager.onTaskCompleted(requestId)
        }
    }
    
    func getTaskInfo(requestId: String) async throws -> TaskInfo? {
        guard var taskInfo = fileStorage.loadTaskInfo(requestId: requestId) else {
            return nil
        }
        
        // If status is IN_PROGRESS, check actual state
        if taskInfo.status == .inProgress {
            // Check if response file exists
            if let responseTaskInfo = fileStorage.loadTaskResponse(requestId: requestId),
               let responseJson = responseTaskInfo.responseJson {
                // Response file exists – request is completed
                if let statusValue = responseJson["status"] as? Int,
                   let status = RequestStatus(rawValue: statusValue),
                   status != .inProgress {
                    // Update status
                    try? fileStorage.saveStatus(requestId: requestId, status: status)
                    // Notify queue that the task is completed
                    await queueManager.onTaskCompleted(requestId)
                    taskInfo = TaskInfo(
                        id: taskInfo.id,
                        status: status,
                        path: taskInfo.path,
                        registrationDate: taskInfo.registrationDate,
                        responseJson: taskInfo.responseJson
                    )
                }
            } else {
                // Check whether the task is in the queue (not started yet)
                if await queueManager.isTaskQueued(requestId) {
                    // Task is in the queue; IN_PROGRESS status is correct
                    return taskInfo
                }
                
                // If the task is neither in the queue nor active, it might be lost
                if await !queueManager.isTaskPendingOrActive(requestId) {
                    // Add it back to the queue
                    _ = await queueManager.enqueue(requestId)
                }
            }
        }
        
        return taskInfo
    }
    
    func getTaskResponse(requestId: String) async throws -> TaskInfo? {
        let response = fileStorage.loadTaskResponse(requestId: requestId)
        
        // If we received a response, notify the queue that the task is completed
        if response?.responseJson != nil {
            await queueManager.onTaskCompleted(requestId)
        }
        
        return response
    }
    
    func cancelTask(requestId: String) async throws -> Bool? {
        guard fileStorage.taskExists(requestId: requestId) else {
            return nil
        }
        
        do {
            // Remove from queue if the task is there
            _ = await queueManager.removeFromQueue(requestId)
            // Cancel in URLSession if the task is there
            urlSession.cancelTask(requestId: requestId)
            // Notify queue that the task is completed
            await queueManager.onTaskCompleted(requestId)
            try fileStorage.saveStatus(requestId: requestId, status: .failed)
            return true
        } catch {
            return false
        }
    }
    
    func deleteTask(requestId: String) async throws -> Bool? {
        guard fileStorage.taskExists(requestId: requestId) else {
            return nil
        }
        
        do {
            // Remove from queue
            _ = await queueManager.removeFromQueue(requestId)
            // Notify queue that the task is completed (in case it was active)
            await queueManager.onTaskCompleted(requestId)
            urlSession.deleteTask(requestId: requestId)
            let deleted = fileStorage.deleteTaskFiles(requestId: requestId)
            return deleted
        } catch {
            return false
        }
    }
    
    func getPendingTasks() async throws -> [PendingTask] {
        // Get all tasks from the file system
        let allTaskIds = fileStorage.getAllTaskIds()
        var pendingTasks: [PendingTask] = []
        
        for requestId in allTaskIds {
            // Check that the task is pending or running
            if let taskInfo = try await getTaskInfo(requestId: requestId),
               taskInfo.status == .inProgress {
                // Check that there is no response yet (task is not completed)
                if let responseInfo = fileStorage.loadTaskResponse(requestId: requestId),
                   responseInfo.responseJson == nil {
                    pendingTasks.append(PendingTask(
                        requestId: requestId,
                        registrationDate: taskInfo.registrationDate
                    ))
                } else if fileStorage.loadTaskResponse(requestId: requestId) == nil {
                    pendingTasks.append(PendingTask(
                        requestId: requestId,
                        registrationDate: taskInfo.registrationDate
                    ))
                }
            }
        }
        
        return pendingTasks
    }
    
    func cancelAllTasks() async throws -> Int {
        // Clear our queue
        let queueCleared = await queueManager.clearAll()
        // Cancel active tasks in URLSession
        _ = urlSession.cancelAllTasks()
        return queueCleared
    }
    
    // MARK: - Queue Management Methods
    
    /// Gets queue statistics
    func getQueueStats() async -> QueueStats {
        return await queueManager.getQueueStats()
    }
    
    /// Sets the maximum number of concurrent tasks
    func setMaxConcurrentTasks(_ count: Int) async {
        await queueManager.setMaxConcurrentTasks(count)
    }
    
    /// Sets the maximum queue size
    func setMaxQueueSize(_ size: Int) async {
        await queueManager.setMaxQueueSize(size)
    }
    
    /// Forces queue processing
    func processQueue() async {
        await queueManager.processQueue()
    }
}

File: android/src/main/kotlin/com/tsitser/background_http_plugin/data/repository/TaskRepositoryImpl.kt
package com.tsitser.background_http_plugin.data.repository

import android.content.Context
import com.tsitser.background_http_plugin.data.datasource.FileStorageDataSource
import com.tsitser.background_http_plugin.data.datasource.TaskQueueManager
import com.tsitser.background_http_plugin.data.datasource.WorkManagerDataSource
import com.tsitser.background_http_plugin.data.mapper.RequestMapper
import com.tsitser.background_http_plugin.domain.entity.HttpRequest
import com.tsitser.background_http_plugin.domain.entity.RequestStatus
import com.tsitser.background_http_plugin.domain.entity.TaskInfo
import com.tsitser.background_http_plugin.domain.repository.PendingTask
import com.tsitser.background_http_plugin.domain.repository.TaskRepository
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.UUID

/**
 * Repository implementation for working with HTTP request tasks.
 *
 * Uses TaskQueueManager to manage the task queue,
 * which helps avoid hangs when registering a large number of tasks.
 */
class TaskRepositoryImpl(
    private val context: Context
) : TaskRepository {

    private val fileStorage = FileStorageDataSource(context)
    private val workManager = WorkManagerDataSource(context)
    private val queueManager = TaskQueueManager.getInstance(context)
    
    // Mutex to protect from race conditions when creating requests with the same requestId
    private val createTaskMutex = Mutex()
    // Set to track requests that are currently being created
    private val creatingTasks = mutableSetOf<String>()

    override suspend fun createTask(request: HttpRequest): TaskInfo {
        val requestId = request.requestId ?: RequestMapper.generateRequestId()
        
        // Race-condition protection: use Mutex for synchronization
        return createTaskMutex.withLock {
            // Check whether a request with such requestId is already being created
            if (creatingTasks.contains(requestId)) {
                // Request is already being created in another thread; wait and return the existing one
                // Small delay to let creation finish in the other thread
                kotlinx.coroutines.delay(100)
                return@withLock getTaskInfo(requestId) 
                    ?: throw IllegalStateException("Failed to get task info after creation attempt")
            }
            
            // Optimization: Fast check if task exists before loading full info
            // This avoids expensive file read for new tasks
            if (fileStorage.taskExists(requestId)) {
                // Task exists, load full info
                val existingTask = fileStorage.loadTaskInfo(requestId)
                if (existingTask != null) {
                    // Request already exists
                    // Check whether it is in the queue or active
                    if (queueManager.isTaskPendingOrActive(requestId)) {
                        // Task is already in the queue or running
                        return@withLock existingTask
                    }
                    
                    // Check state in WorkManager
                    val workState = workManager.getWorkState(requestId, forceRefresh = false)
                    when (workState) {
                        WorkManagerDataSource.WorkStateResult.IN_PROGRESS,
                        WorkManagerDataSource.WorkStateResult.SUCCEEDED,
                        WorkManagerDataSource.WorkStateResult.FAILED -> {
                            // Request is already running or completed; return existing
                            return@withLock existingTask
                        }
                        WorkManagerDataSource.WorkStateResult.NOT_FOUND -> {
                            // Request exists in file storage, but not in WorkManager and not in the queue
                            // Enqueue it for execution
                            queueManager.enqueue(requestId)
                            return@withLock existingTask
                        }
                    }
                }
            }
            // Optimization: For new tasks (taskExists == false), skip WorkManager check
            // New tasks don't exist in WorkManager, so checking is unnecessary
            
            // Mark that the request is being created
            creatingTasks.add(requestId)
            
            try {
                val registrationDate = System.currentTimeMillis()
            
                // Save request to file
                val taskInfo = fileStorage.saveRequest(request, requestId, registrationDate)
            
                // Add task to the queue (instead of starting it directly in WorkManager)
                // TaskQueueManager itself will decide when to start the task
                queueManager.enqueue(requestId)
            
                return@withLock taskInfo
            } finally {
                // Remove from the set of requests being created
                creatingTasks.remove(requestId)
            }
        }
    }

    override suspend fun getTaskInfo(requestId: String): TaskInfo? {
        // Load task information from file
        var taskInfo = fileStorage.loadTaskInfo(requestId) ?: return null
        
        // If status is IN_PROGRESS, check actual state
        if (taskInfo.status == RequestStatus.IN_PROGRESS) {
            // First check if the response file exists – this is faster than querying WorkManager
            // If the response file exists, the request is completed
            val responseTaskInfo = fileStorage.loadTaskResponse(requestId)
            if (responseTaskInfo != null && responseTaskInfo.responseJson != null) {
                // Response file exists – request is completed
                // Update status from response
                val responseStatus = responseTaskInfo.responseJson["status"] as? Int
                if (responseStatus != null) {
                    val status = enumValues<RequestStatus>().firstOrNull { it.value == responseStatus }
                    if (status != null && status != RequestStatus.IN_PROGRESS) {
                        // Update status in status file for consistency
                        fileStorage.saveStatus(requestId, status)
                        // Notify queue that the task is completed
                        queueManager.onTaskCompleted(requestId)
                        // Return updated information
                        taskInfo = taskInfo.copy(status = status)
                    }
                }
            } else {
                // Check whether the task is in the queue (not started yet)
                if (queueManager.isTaskQueued(requestId)) {
                    // Task is in the queue; IN_PROGRESS status is correct
                    return taskInfo
                }
                
                // If there is no response file, check state in WorkManager.
                // This helps determine whether the task is finished but the response file is not yet written.
                // Use caching for optimization when there are many requests.
                val workState = workManager.getWorkState(requestId, forceRefresh = false)
                when (workState) {
                    WorkManagerDataSource.WorkStateResult.SUCCEEDED -> {
                        // WorkManager reports that the task has completed successfully,
                        // but the response file has not yet been created – it may still be writing.
                        // In this case we keep IN_PROGRESS so that on the next call
                        // the response file will be available and status will be updated from it.
                    }
                    WorkManagerDataSource.WorkStateResult.FAILED -> {
                        // WorkManager reports that the task has failed.
                        // Update status.
                        fileStorage.saveStatus(requestId, RequestStatus.FAILED)
                        // Notify queue that the task is completed
                        queueManager.onTaskCompleted(requestId)
                        taskInfo = taskInfo.copy(status = RequestStatus.FAILED)
                    }
                    WorkManagerDataSource.WorkStateResult.IN_PROGRESS -> {
                        // Task is indeed running in WorkManager.
                        // IN_PROGRESS status in the file is correct; do nothing.
                    }
                    WorkManagerDataSource.WorkStateResult.NOT_FOUND -> {
                        // Task not found in WorkManager and not in queue.
                        // It may have been lost – add it back to the queue.
                        if (!queueManager.isTaskPendingOrActive(requestId)) {
                            queueManager.enqueue(requestId)
                        }
                    }
                }
            }
        }
        
        return taskInfo
    }

    override suspend fun getTaskResponse(requestId: String): TaskInfo? {
        val response = fileStorage.loadTaskResponse(requestId)
        
        // If we received a response, notify the queue that the task is completed
        if (response?.responseJson != null) {
            queueManager.onTaskCompleted(requestId)
        }
        
        return response
    }

    override suspend fun cancelTask(requestId: String): Boolean? {
        if (!fileStorage.taskExists(requestId)) {
            return null
        }
        
        return try {
            // Remove from queue if the task is there
            queueManager.removeFromQueue(requestId)
            // Cancel in WorkManager if the task is there
            workManager.cancelRequest(requestId)
            // Notify queue that the task is completed
            queueManager.onTaskCompleted(requestId)
            fileStorage.saveStatus(requestId, RequestStatus.FAILED)
            true
        } catch (e: Exception) {
            false
        }
    }

    override suspend fun deleteTask(requestId: String): Boolean? {
        if (!fileStorage.taskExists(requestId)) {
            return null
        }
        
        return try {
            // Remove from queue
            queueManager.removeFromQueue(requestId)
            // Notify queue that the task is completed (in case it was active)
            queueManager.onTaskCompleted(requestId)
            workManager.deleteRequest(requestId)
            val deleted = fileStorage.deleteTaskFiles(requestId)
            deleted
        } catch (e: Exception) {
            false
        }
    }

    override suspend fun getPendingTasks(): List<PendingTask> {
        val pendingTasks = mutableListOf<PendingTask>()
        
        // Get all tasks from the file system
        val allTaskIds = fileStorage.getAllTaskIds()
        
        for (requestId in allTaskIds) {
            // Check that the status in file is IN_PROGRESS (not completed)
            val taskInfo = fileStorage.loadTaskInfo(requestId)
            if (taskInfo != null && taskInfo.status == RequestStatus.IN_PROGRESS) {
                // Check that there is no response yet (task is not completed)
                val response = fileStorage.loadTaskResponse(requestId)
                if (response == null || response.responseJson == null) {
                    // Add task to pending list
                    pendingTasks.add(PendingTask(
                        requestId = requestId,
                        registrationDate = taskInfo.registrationDate
                    ))
                }
            }
        }
        
        return pendingTasks
    }

    override suspend fun cancelAllTasks(): Int {
        // Clear our queue and WorkManager
        val queueCleared = queueManager.clearAll()
        workManager.cancelAllTasks()
        return queueCleared
    }
    
    /**
     * Gets queue statistics.
     */
    fun getQueueStats(): TaskQueueManager.QueueStats {
        return queueManager.getQueueStats()
    }
    
    /**
     * Sets the maximum number of concurrent tasks.
     */
    suspend fun setMaxConcurrentTasks(count: Int) {
        queueManager.setMaxConcurrentTasks(count)
    }
    
    /**
     * Sets the maximum queue size.
     */
    fun setMaxQueueSize(size: Int) {
        queueManager.setMaxQueueSize(size)
    }
    
    /**
     * Synchronizes the queue state.
     */
    suspend fun syncQueueState() {
        queueManager.syncQueueState(fileStorage)
    }
    
    /**
     * Forces queue processing.
     */
    suspend fun processQueue() {
        queueManager.processQueue()
    }
}
